src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/OpenLdapUnsafeMethods.cs

```diff
@@ -0,0 +1,192 @@
+        public static extern int ldap_delete_ext([In] ConnectionHandle ldapHandle, string dn, IntPtr servercontrol, IntPtr clientcontrol, ref int messageNumber);
+
+        [DllImport(_openldapdll, EntryPoint = "ldap_rename", CharSet = CharSet.Ansi)]
+        public static extern int ldap_rename([In] ConnectionHandle ldapHandle, string dn, string newRdn, string newParentDn, int deleteOldRdn, IntPtr servercontrol, IntPtr clientcontrol, ref int messageNumber);
+
+        [DllImport(_openldapdll, EntryPoint = "ldap_compare_ext", CharSet = CharSet.Ansi)]
+        public static extern int ldap_compare([In] ConnectionHandle ldapHandle, string dn, string attributeName, berval binaryValue, IntPtr servercontrol, IntPtr clientcontrol, ref int messageNumber);
+
+        // Basic Encoding Rules (BER) native functions
+        [DllImport(_openldapdll, EntryPoint = "ber_alloc_t", CharSet = CharSet.Ansi)]
+        public static extern IntPtr ber_alloc(int option);
+
+        [DllImport(_openldapdll, EntryPoint = "ber_init", CharSet = CharSet.Ansi)]
+        public static extern IntPtr ber_init(berval value);
+
+        [DllImport(_openldapdll, EntryPoint = "ber_free", CharSet = CharSet.Ansi)]
+        public static extern IntPtr ber_free([In] IntPtr berelement, int option);
+
+        [DllImport(_openldapdll, EntryPoint = "ber_printf", CharSet = CharSet.Ansi)]
+        public static extern int ber_printf_emptyarg(BerSafeHandle berElement, string format);
+
+        [DllImport(_openldapdll, EntryPoint = "ber_printf", CharSet = CharSet.Ansi)]
+        public static extern int ber_printf_int(BerSafeHandle berElement, string format, int value);
+
+        [DllImport(_openldapdll, EntryPoint = "ber_printf", CharSet = CharSet.Ansi)]
+        public static extern int ber_printf_bytearray(BerSafeHandle berElement, string format, HGlobalMemHandle value, int length);
+
+        [DllImport(_openldapdll, EntryPoint = "ber_printf", CharSet = CharSet.Ansi)]
+        public static extern int ber_printf_berarray(BerSafeHandle berElement, string format, IntPtr value);
+
+        [DllImport(_openldapdll, EntryPoint = "ber_flatten", CharSet = CharSet.Ansi)]
+        public static extern int ber_flatten(BerSafeHandle berElement, ref IntPtr value);
+
+        [DllImport(_openldapdll, EntryPoint = "ber_bvfree", CharSet = CharSet.Ansi)]
+        public static extern int ber_bvfree(IntPtr value);
+
+        [DllImport(_openldapdll, EntryPoint = "ber_bvecfree", CharSet = CharSet.Ansi)]
+        public static extern int ber_bvecfree(IntPtr value);
+
+        [DllImport(_openldapdll, EntryPoint = "ber_scanf", CharSet = CharSet.Ansi)]
+        public static extern int ber_scanf(BerSafeHandle berElement, string format);
+
+        [DllImport(_openldapdll, EntryPoint = "ber_scanf", CharSet = CharSet.Ansi)]
+        public static extern int ber_scanf_int(BerSafeHandle berElement, string format, ref int value);
```

Comment:
We don't have a rule here, but it bugs the heck out of me:

Please make all `public` methods in `internal` classes be `internal`, unless they're required to be `public` for interface matching.  It stops API Review alarm bells from going off in my head.

Reply:
Perhaps we should create and ship an analyzer for that? (Off by default, on for us)?

Reply:
Jeremy and I have exact opposite cases of what bugs us ;)

Reply:
Interesting -- when do you use internal methods in internal classes?

Reply:
I generally imagine what the type would look like if I flipped the type's visibility to be public, and mark the visibility on the members accordingly.

Reply:
🤯

Reply:
> I generally imagine what the type would look like if I flipped the type's visibility to be public, and mark the visibility on the members accordingly.

Me, too.  And I don't think we'd want any of these to be public :).

Reply:
Sure we would, _if_ we were making the type public (which we're not). The thought experiment is, given the nature of the type, what are the APIs you want consumers to be able to use without violating the type's design (public) vs minor peeks into its implementation (internal) vs entire implementation (private).

Reply:
ok, given the above discussion I guess for now I will set them as internal given it is just a static class focused on P/Invokes and we don't want to upset @bartonjs 😆